File: 02_changes_net_worth/query.py
import adbc_driver_sqlite.dbapi as dbapi
import polars as pol
import pyarrow as pa
import arrowkit

def main(table:str):

    conn = dbapi.connect("file:/content/my_SQLite.db?mode=ro")
    name = table.upper()
    if (name == "TRANSACTIONS_02"):

        try:

            arrow_transactions = arrowkit.get_ArrowTable(conn, table)
            print(f'RETURNING TABLE TRANSACTIONS_02 TABLE FROM DATABASE:\n{arrow_transactions.schema}')
            print(":" * 40)
            transactions = pol.from_arrow(arrow_transactions).lazy()

            sample = transactions.head(5)
            print(f'TRANSACTIONS TABLE (POLARS) -> SAMPLE:\n{sample.collect()}')
            _type = (transactions.unpivot(on = ['SENDER', 'RECEIVER'],
                                         index = 'AMOUNT',
                                         variable_name = 'TYPE',
                                         value_name = 'USER_ID'
                                 )
                                .with_columns(pol.when(pol.col('TYPE') == 'SENDER')
                                                 .then(pol.col('AMOUNT') * -1)
                                                 .otherwise(pol.col('AMOUNT'))
                                                 .alias('AMOUNT')
                                 )
                                .group_by(pol.col('USER_ID'))
                                .agg(pol.col('AMOUNT').sum())
                                .sort(by = 'AMOUNT', 
                                     descending = True
                                 )
            )
            print(":" * 40)
            print(f'NET CHANGES MADE BY EACH USER (POLARS):\n{_type.collect()}')

            query = """
                    WITH 
                        SENDERS AS (
                            SELECT
                                SENDER,
                                SUM(AMOUNT) AS SENDED
                            FROM 
                                TRANSACTIONS_02
                            GROUP BY
                                SENDER),
                        RECEIVERS AS (
                            SELECT
                                RECEIVER,
                                SUM(AMOUNT) AS RECEIVED
                            FROM
                                TRANSACTIONS_02
                            GROUP BY
                                RECEIVER) 
                    SELECT
                        COALESCE(S.SENDER, R.RECEIVER) AS USER_ID,
                        COALESCE(R.RECEIVED, 0) - COALESCE(S.SENDED, 0) AS NET_CHANGE
                    FROM
                        RECEIVERS R 
                    FULL JOIN 
                        SENDERS S 
                    ON 
                        (R.RECEIVER = S.SENDER)
                    ORDER BY 
                        2 DESC
            """
            result = arrowkit.get_ArrowTable(conn, query)
            df = result.to_pandas()
            print(":" * 40)
            print(f'NET CHANGES MADE BY EACH USER (DuckDB):\n{result}')
            print(f'<*pandas visualization*>\n{df}')

        finally:

            conn.close()

    else:

        print(f'TABLE {table} NOT AVAILABLE')

File: 02_changes_net_worth/query.jl
const DB_PATH = "my_SQLite.db"
using .MyDataBase, DataFrames, Arrow, SQLite, DuckDB, .SQLiteArrowKit, PrettyTables 

MyDataBase.main()
#=
**********************************************
=#
function print_DuckTable(cursor::DuckDB.QueryResult)

    pretty_table(
                 cursor;
                 tf = tf_unicode,
                 hlines = [:begin, 1],
                 vlines = [0, :end]
    )
end
#=
**********************************************
=#
function main(args = ARGS)

    db = SQLite.DB(DB_PATH)
    name = uppercase(args)

    if is_available(db, args) && (name == "TRANSACTIONS_02")

        duck = nothing

        try

            duck = DBInterface.connect(DuckDB.DB, ":memory:")
            arrow_transactions = get_ArrowTable(db, args)
            println("RETURNING TABLE TRANSACTIONS_02 FROM DATABASE:\n$arrow_transactions")

            DuckDB.register_data_frame(duck, arrow_transactions, "arrow_transactions")
            sample = """
            SELECT 
                * 
            FROM 
                'arrow_transactions'
            USING 
                SAMPLE 50% (bernoulli)
            """
            duck_sample = DBInterface.execute(duck, sample)
            println("\n", "*"^40)
            println("TRANSACTIONS TABLE (DuckDB) -> SAMPLE:")
            print_DuckTable(duck_sample)

            query = """
            WITH SENDERS AS (
                SELECT
                    SENDER,
                    SUM(AMOUNT) AS SENDED
                FROM
                    'arrow_transactions'
                GROUP BY
                    SENDER),
            RECEIVERS AS (
                SELECT
                    RECEIVER,
                    SUM(AMOUNT) AS RECEIVED
                FROM
                    'arrow_transactions'
                GROUP BY
                    RECEIVER) SELECT
                                  COALESCE(S.SENDER, R.RECEIVER) AS USER_ID,
                                  COALESCE(R.RECEIVED, 0) - COALESCE(S.SENDED, 0) AS NET_CHANGE
                              FROM
                                  RECEIVERS R
                              FULL JOIN SENDERS S ON (R.RECEIVER = S.SENDER)
                              ORDER BY 2 DESC
            """
            duck_query = DBInterface.execute(duck, query)
            println("\n", "*"^40)
            println("NET CHANGES MADE BY EACH USER (DuckDB):")
            print_DuckTable(duck_query)
            df_transactions = arrow_transactions |> DataFrame

            df = (rename!(stack(select(df_transactions,
                                       :AMOUNT,
                                       :SENDER,
                                       :RECEIVER
                                      ),
                                2:3
                               ),
                          :variable => :TYPE,
                          :value => :USER_ID
                         )
            )
            transform!(df, [:TYPE, :AMOUNT] => ByRow((x, y) -> x == "SENDER" ? -1 * y : y) => :AMOUNT)

            result = (sort(combine(groupby(df,
                                           :USER_ID
                                          ),
                                   :AMOUNT => sum
                                  ),
                           :AMOUNT_sum,
                           rev = true
                          )
            )
            println("\n", "*"^40)
            println("NET CHANGES MADE BY EACH USER (DataFrames.jl):\n$result")

        finally

            DBInterface.close!(duck)

        end

    else

        println(" TABLE $args not AVAILABLE")

    end

end
#=
**********************************************
=#
if Base.@isdefined(PROGRAM_FILE) &&
   abspath(PROGRAM_FILE) == abspath(@__FILE__)

    main()

end
